libs/gl-client/src/signer/mod.rs

```diff
@@ -761,16 +766,145 @@ impl Signer {
     ) -> Result<(), anyhow::Error> {
         let scheduler = self.init_scheduler(scheduler_uri).await?;
         tokio::select! {
-            run_forever_inner_res = self.run_forever_inner(scheduler) => {
+            run_forever_inner_res = self.run_forever_inner(scheduler.clone()) => {
                 error!("Inner signer loop exited unexpectedly: {run_forever_inner_res:?}");
             },
+            run_forever_scheduler_res = self.run_forever_scheduler(scheduler) => {
+                error!("Scheduler signer loop exited unexpectedly: {run_forever_scheduler_res:?}")
+            }
             _ = shutdown.recv() => debug!("Received the signal to exit the signer loop")
         };
 
         info!("Exiting the signer loop");
         Ok(())
     }
 
+    async fn run_forever_scheduler(
+        &self,
+        mut scheduler: SchedulerClient<tonic::transport::Channel>,
+    ) -> Result<(), anyhow::Error> {
+        let (sender, rx) = mpsc::channel(1);
+        let outbound = ReceiverStream::new(rx);
+        // let inbound_future = scheduler.signer_requests_stream(outbound);
+
+        let mut stream = scheduler
+            .signer_requests_stream(outbound)
+            .await?
+            .into_inner();
+
+        trace!("Starting to stream signer requests from scheduler");
+
+        loop {
+            match stream.message().await {
+                Ok(Some(msg)) => {
+                    let req_id = msg.request_id;
+                    trace!("Processing scheduler request {}", req_id);
+                    match msg.request {
+                        Some(signer_request::Request::ApprovePairing(req)) => {
+                            self.process_pairing_approval(req_id, req, sender.clone())
+                                .await
+                        }
+                        None => {
+                            debug!("Received an empty signing request");
+                        }
+                    };
+                }
+                Ok(None) => {
+                    debug!("End of stream, this should not happen by the server");
+                    return Err(anyhow!("Scheduler closed the stream"));
+                }
+                Err(e) => {
+                    debug!("Got an error from the scheduler {}", e);
+                    return Err(anyhow!("Scheduler stream error {e:?}"));
+                }
+            };
+        }
+    }
+
+    async fn process_pairing_approval(
+        &self,
+        req_id: u32,
+        req: ApprovePairingRequest,
+        stream: mpsc::Sender<SignerResponse>,
+    ) -> () {
+        let mut data = vec![];
+        data.put(req.session_id.as_bytes());
+        data.put_u64(req.timestamp);
```

Comment:
Should the timestamp be checked in this method to bound its staleness like what was mentioned here?:

https://github.com/Blockstream/greenlight/pull/281/files#diff-95a68fef05d112753f7fbc0859cb6df74b6df1b75994ceb8937170b228651c88R515

Reply:
As the timestamp is part of the signature we also need to add it to the data in order to verify that the signature is correct.